### ISSUE TITLE
How to remove Primary Instance?

### ISSUE THREAD

**Issue by arthurvanduynhoven** [jhuckaby/Cronicle#393]
## Summary

I am using a Multi-Server Cluster (2-5 instances) with no workers with S3 backend. I want to remove the primary instance so another can become primary (rolling upgrades).

I tried to just remove the master with the UI and it says I can't (as it's master).

I then tried to just terminate the master instance, but then no other servers will become master. They eventually say they are no longer eligible (see below logs).

## Steps to reproduce the problem

Spin up a Multi-Server cluster in AWS ECS Fargate. 

## Your Setup

AWS ECS Fargate

### Operating system and version?

Docker NODE LTS container

### Node.js version?

latest LTS

### Cronicle software version?

Latest

### Are you using a multi-server setup, or just a single server?
Multi Server

### Are you using the filesystem as back-end storage, or S3/Couchbase?
S3

### Can you reproduce the crash consistently?
Yes in AWS ECS Fargate

## Log Excerpts

`
[1616508345.922][2021-03-23 14:05:45][ip-10-22-76-13.eu-central-1.compute.internal][117][Cronicle][debug][3][Master server is: ip-10-22-76-186.eu-central-1.compute.internal][]

[1616508345.922][2021-03-23 14:05:45][ip-10-22-76-13.eu-central-1.compute.internal][117][Cronicle][debug][3][We are becoming a slave server][]

[1616508345.922][2021-03-23 14:05:45][ip-10-22-76-13.eu-central-1.compute.internal][117][Cronicle][debug][4][Looking for leftover job JSON files: logs/jobs/*.json][]

[1616508345.925][2021-03-23 14:05:45][ip-10-22-76-13.eu-central-1.compute.internal][117][Cronicle][debug][4][Server is eligible to become master (Primary Group)][]

[1616508345.927][2021-03-23 14:05:45][ip-10-22-76-13.eu-central-1.compute.internal][117][Cronicle][debug][4][Looking for leftover job logs: logs/jobs/*.log][]

[1616508566.446][2021-03-23 14:09:26][ip-10-22-76-13.eu-central-1.compute.internal][117][Cronicle][debug][5][Socket.io client disconnected: u1MuWFdH7V-obGmCAAAA (IP: ::ffff:10.22.76.186)][]

[1616508583.395][2021-03-23 14:09:43][ip-10-22-76-13.eu-central-1.compute.internal][117][Cronicle][debug][3][No master ping received within 30 seconds, choosing new master][]

[1616508583.397][2021-03-23 14:09:43][ip-10-22-76-13.eu-central-1.compute.internal][117][Cronicle][debug][4][We are no longer eligible for master, going into idle mode][]
`

**Comment by mikeTWC1984** [comment]
Looks like native failover only works with 2 manager nodes. I tried 3 manager cluster, and indeed if I kill current leader, 2 other nodes get confused (UI gets unstable), but after killing one more, the last node become a manager.
From the other side, I think if you use some sort of container management  then 1 manager / N workers seemed to be more practical solution.

**Comment by dheise-rheinenergie-trading-com** [comment]
I use a setup with 2 master servers and 1 worker (the masters are also working ;-) ). If I stop my primary master, my UI is unstable only for the time defined in the config under "master_ping_freq". After waiting the defined time, my backup server overtakes the master role and the system becomes stable again.
So I normally I wait in the server overview till a new minute (watching the cronicle clock in the upper right) begins and my current master finished all his running events. Then I can savely stop him and 15 seconds later (my defined master_ping_freq), the backup is the new master. This all happens normally in under 1 Minute, so that I don't miss any execution.

**Comment by mikeTWC1984** [comment]
@dheise-rheinenergie-trading-com  you are right, it take some time to become a manager. Seems like "unstable" state is normal for backup nodes, it's not caused by loosing leader. I tried 3 and 5 manager node clusters and failover occurred within a minute (using default configs). Although I'm using filesystem storage and those nodes do not share log directory. Will try to emulate s3 storage later

**Comment by arthurvanduynhoven** [comment]
Thanks for the information and help.

I am curious to see your findings when using S3 storage.

In the meantime I will limit to only have a max of 2 nodes in the primary group and will see if that will fulfil my needs.

**Comment by mikeTWC1984** [comment]
I tested a cluster of 3 manager nodes using s3 (minio though). Everything works same way - 2nd node become a leader after some time.
As far as I can see from source code, you are getting "We are no longer eligible for master" message because that server name doesn't match the manager group regex. 
https://github.com/jhuckaby/Cronicle/blob/bba7b01579d56b1354392d1f3e710cec3270093a/lib/engine.js#L567
How does your primary group regex look like?

**Comment by arthurvanduynhoven** [comment]
I set the regex the same as the all servers .+

In the logs you see the node specify its eligible to become master, but after the master node is removed it then says it’s no longer eligible. So not sure exactly what is going on with it.

**Comment by mikeTWC1984** [comment]
OK. My only guess then is that your other nodes do not have access to the storage. E.g. because of aws-sdk package is not installed or nodes using different configs.
But as I mentioned before you probably don't really need multi-master setup. As far as I understand fargate is some sort container orchestration thing, so it will take care or getting your main node up and running. All other nodes could be simple workers, without setting up storage. Btw, aws-sdk package is about 100MB (cronicle itself I think is about 50MB). You can also set config master=1 or env variable CRONICLE_master=1 (if using single manager), so after restart it will be up almost right away and  base image update would take just several seconds.

**Comment by jhuckaby** [comment]
@KursatKaya Oh wow, that's a great find.  Yes, only the master node should have caching enabled.  I'm glad you found a workaround, but this is totally a bug.  Non-master nodes should automatically disable caching, and then it should re-enable itself if the backup node becomes master.  I'll work on this for the next release.  Thanks!
